### PR TITLE
All the branch instructions

### LINKED ISSUE #27: BCS
Fully assemble, and simulate it.

### LINKED ISSUE #28: BGE
Fully assemble, and simulate it.

### LINKED ISSUE #29: BLT
Fully assemble, and simulate it.

### LINKED ISSUE #30: BVC
Fully assemble, and simulate it.

### LINKED ISSUE #31: BRA
Fully assemble, and simulate it.

### LINKED ISSUE #90: Add BRA absolute offsets
This is in the format `BRA *<+/-><number>`

### PR DESCRIPTION
**Pull request by pupper68k**
Tasks:
 - base class for Branch Instructions (DONE)
 - edits to assembler.py to pass in the address of the branch instruction (DONE)
 - overloads for each 16 branch instructions (DONE)
 - tests for each branch instruction (DONE)

This PR will track the changes made to support branch instructions.

Fix #27 
Fix #28 
Fix #29 
Fix #30 
Fix #31 
Fix #90

### CODE REVIEW

**Comment by pupper68k** [comment]
Edits to assembler have been done, Branch instructions now have all they need to execute properly.
Code review welcome at this point.

Overload for BRA done, now testing.

File: easier68k/core/opcodes/Bcc.py [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+        Gets what the end length of this command will be in memory
+        :param command: The text of the command itself (e.g. "LEA", "MOVE.B", etc.)
+        :param parameters: The parameters after the command
+        :return: The length of the bytes in memory in words
+        """
+        
+        # opword
+        length = 2
+        
+        # BIG NOTE: I am explicitly ignoring size codes because there are no defined flippy boys
```
Comment (Chris-Johnston) [inline easier68k/core/opcodes/Bcc.py]: Do we have to worry about when the 8-bit displacement is 0 or `0xFF` here, or is that not an issue?

See page 130: https://www.nxp.com/files-static/archives/doc/ref_manual/M68000PRM.pdf

File: easier68k/core/opcodes/Bcc.py [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+        :param command: The command itself (e.g. 'MOVE.B', 'LEA', etc.)
+        :param parameters: The parameters after the command (such as the source and destination of a move)
+        :return: Whether the given command is valid and a list of issues/warnings encountered
+        """
+        
+        passes = True
+        errors = []
+        parameters = parameters.split(',')
+        
+        # See the BIG HORKIN' NOTE from get_word_length()
```
Comment (Chris-Johnston) [inline easier68k/core/opcodes/Bcc.py]: Comments are funny, but I think I'd prefer if they were in a more neutral tone

File: easier68k/core/opcodes/Bcc.py [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+        # check param length
+        if len(parameters) >2:
+            passes = False
+            errors.append("Too many operands for Branch Operation (TODO: list them here)")
```
Comment (Chris-Johnston) [inline easier68k/core/opcodes/Bcc.py]:
```suggestion
            errors.append(f"Too many operands for Branch Operation: " + ', '.join(parameters))
```

File: easier68k/core/opcodes/Bcc.py [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+        
+        # check offset dest validity (odd read)
+        operand = parse_assembly_parameter(parameters[0].strip())
+        if operand.data % 2 == 1:
+            passes = False
+            errors.append("Reading word at {d} will cause crash".format(d=operand.data))
```
Comment (Chris-Johnston) [inline easier68k/core/opcodes/Bcc.py]: if only we could target >3.6 we could use f strings here

File: easier68k/core/opcodes/Bcc.py [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+    def disassemble_instruction(cls, data: bytes) -> Opcode:
+        """
+        Parses some raw data into an instance of the opcode class
+
+        :param data: The data used to convert into an opcode instance
+        :return: The constructed instance or none if there was an error and
+            the amount of data in words that was used (e.g. extra for immediate
+            data) or 0 for not a match
+        """
+        raise Exception("Branch Operation did not override parent class disassemble_instruction stub!")
```
Comment (Chris-Johnston) [inline easier68k/core/opcodes/Bcc.py]: I think we should raise `NotImplementedError` instead

File: easier68k/core/opcodes/Bcc.py [hunk 6]
```diff
@@ -0,0 +1,358 @@
...
+    def from_str(cls, command: str, parameters: str):
+        """
+        Parses a command from text
+
+        :param command: The command itself (e.g. 'MOVE.B', 'LEA', etc.)
+        :param parameters: The parameters after the command (such as the source and destination of a move)
+        :param address: Address of this operation 
+        :return: The parsed command
+        """
+        raise Exception("Branch Operation did not override parent class from_str stub!")
```
Comment (Chris-Johnston) [inline easier68k/core/opcodes/Bcc.py]: same thing from above

File: easier68k/core/opcodes/Bcc.py (hunk 4 above)
Comment (pupper68k) [reply easier68k/core/opcodes/Bcc.py]: On one hand most distributions of linux ship 3.6
On the other, current stable is 3.7.2 iirc

File: easier68k/core/opcodes/Bcc.py (hunk 1 above)
Comment (pupper68k) [reply easier68k/core/opcodes/Bcc.py]: To be sure, I compiled the following code in Easy68k and got "60FE" back out of it
```
L    BRA L
```
I will just add a line to the constructor of the branch_code base class that sets displacement to -1 when it is 0.

Comment (pupper68k) [reply easier68k/core/opcodes/Bcc.py]: "Note: A branch to the immediately following instruction automatically
uses   the   16-bit   displacement   format   because   the   8-bit
displacement field contains $00 (zero offset)." 

0xFF on the other hand, is an illegal access exception, as you would instruct the sim to pull a word at an odd byte-address, causing the sim to need two read operations to get the word.

Two corner cases which definitely need to be added to my code.

**Comment by pupper68k** [comment]
So theres tests and overloads for all of the branch operations.
Let me know if theres any changes that need to be made to this :)

File: easier68k/core/opcodes/Bcc.py [hunk 7]
```diff
@@ -364,4 +364,403 @@ def disassemble_instruction(cls, data: bytes):
...
+    The BCS opcode compiles down to 65xx xxxx xxxx
+    The conditional here is C
+    """
+
+    cond_code = '\x05'
+
+    def conditional(self, simulator: M68K):
+        c = simulator.get_condition_status_code(ConditionStatusCode.C)
+
+        if c:
```
Comment (Chris-Johnston) [inline easier68k/core/opcodes/Bcc.py]: So `get_condition_status_code` returns a `bool`. We could just do: `return c` here.

File: easier68k/core/opcodes/Bcc.py [hunk 8]
```diff
@@ -364,4 +364,403 @@ def disassemble_instruction(cls, data: bytes):
...
+    The BNE opcode compiles down to 66xx xxxx xxxx
+    The conditional here is NOT Z
+    """
+
+    cond_code = '\x06'
+
+    def conditional(self, simulator: M68K):
+        z = simulator.get_condition_status_code(ConditionStatusCode.Z)
+
+        if z:
```
Comment (Chris-Johnston) [inline easier68k/core/opcodes/Bcc.py]: Same with all the other cases

File: easier68k/core/opcodes/Bcc.py [hunk 9]
```diff
@@ -364,4 +364,403 @@ def disassemble_instruction(cls, data: bytes):
...
+    The conditional here is N AND V OR NOT N AND NOT V
+    """
+
+    cond_code = '\x0C'
+
+    def conditional(self, simulator: M68K):
+        n = simulator.get_condition_status_code(ConditionStatusCode.N)
+        v = simulator.get_condition_status_code(ConditionStatusCode.V)
+
+        if n and v:
```
Comment (Chris-Johnston) [inline easier68k/core/opcodes/Bcc.py]: Couldn't we simplify this to `return n == v`?

File: easier68k/core/opcodes/Bcc.py [hunk 10]
```diff
@@ -364,4 +364,403 @@ def disassemble_instruction(cls, data: bytes):
...
+    The conditional here is N AND NOT V OR NOT N AND V
+    """
+
+    cond_code = '\x0D'
+
+    def conditional(self, simulator: M68K):
+        n = simulator.get_condition_status_code(ConditionStatusCode.N)
+        v = simulator.get_condition_status_code(ConditionStatusCode.V)
+
+        if n and v:
```
Comment (Chris-Johnston) [inline easier68k/core/opcodes/Bcc.py]: This does not match `N AND NOT V OR NOT N AND V`
Also, we could probably do `return n != v`

File: easier68k/core/opcodes/Bcc.py [hunk 11]
```diff
@@ -364,4 +364,403 @@ def disassemble_instruction(cls, data: bytes):
...
+        v = simulator.get_condition_status_code(ConditionStatusCode.V)
+        z = simulator.get_condition_status_code(ConditionStatusCode.Z)
+
+        if n and v and not z:
+            return True
+
+        if not n and not v and not z:
+            return True
+
+        return False
```
Comment (Chris-Johnston) [inline easier68k/core/opcodes/Bcc.py]: Could be simplified to `return (n == v) and not z`

File: easier68k/core/opcodes/Bcc.py [hunk 12]
```diff
@@ -364,4 +364,403 @@ def disassemble_instruction(cls, data: bytes):
...
+        if z:
+            return True
+
+        if v and not z:
+            return True
+
+        if not n and not v and not z:
+            return True
+
+        return False
```
Comment (Chris-Johnston) [inline easier68k/core/opcodes/Bcc.py]: `return (n != v) and not z`

File: easier68k/core/opcodes/Bcc.py [hunk 13]
```diff
@@ -364,4 +364,403 @@ def disassemble_instruction(cls, data: bytes):
...
+
+    @classmethod
+    def from_str(cls, command: str, parameters: str):
+        return bcc_from_str(command, parameters, BGT)
+
+
+class BLE(branch_code):
+    """
+    The BLE opcode compiles down to 6Fxx xxxx xxxx
+    The conditional here is Z OR V AND NOT Z OR NOT N AND NOT V AND NOT Z
```
Comment (Chris-Johnston) [inline easier68k/core/opcodes/Bcc.py]: We should probably group conditions by parenthesis in this comment, for clarity.

File: tests/easier68k/core/opcodes/test_branches.py [hunk 14]
```diff
@@ -65,7 +66,24 @@ def test_bra_disassemble():
...
     #assert b.address == who knows
 
-# test validity (for valid and invalid invluding odd instructions)
+def test_bra_validation():
+    """
+    Test if BRA is_valid commands work as expected
+    :return:
+    """
+
+    # BRA.is_valid("BRA", "$1000")
```
Comment (Chris-Johnston) [inline tests/easier68k/core/opcodes/test_branches.py]: intentional? we should use these tests.

File: tests/easier68k/core/opcodes/test_branches.py [hunk 15]
```diff
@@ -65,7 +66,24 @@ def test_bra_disassemble():
...
+    #    .L, .W, .B, (odd), (even)
+    pass
+
+def test_bra_length():
+    """
+    Test BRA.get_word_length()
+    :return:
+    """
+
+    # BRA.getWordLength() (.b .w .l)
```
Comment (Chris-Johnston) [inline tests/easier68k/core/opcodes/test_branches.py]: We should also use these tests?

File: easier68k/core/opcodes/Bcc.py (hunk 4 above)
Comment (Chris-Johnston) [reply easier68k/core/opcodes/Bcc.py]: True, but do we want to break 3.5? It's not crucial, so maybe not.

File: easier68k/core/opcodes/Bcc.py (hunk 10 above)
Comment (pupper68k) [reply easier68k/core/opcodes/Bcc.py]: broke the tests so I assume not

File: easier68k/core/opcodes/Bcc.py (hunk 11 above)
Comment (pupper68k) [reply easier68k/core/opcodes/Bcc.py]: not so sure that is true.
The manual does not say that  N & V & ~Z are valid.

File: easier68k/core/opcodes/Bcc.py (hunk 12 above)
Comment (pupper68k) [reply easier68k/core/opcodes/Bcc.py]: ~N & ~V &~Z is not covered by this conditional.

The code I wrote is meant to directly reflect the conditionals in the m68k manual by NXP.
If you would like them more simplified I suppose I could probably bust out a Karnaugh Map for these three variable conditionals. I think this is as concise as It gets though.

File: tests/easier68k/core/opcodes/test_branches.py (hunk 14 above)
Comment (pupper68k) [reply tests/easier68k/core/opcodes/test_branches.py]: intentional. we are using that test.

File: tests/easier68k/core/opcodes/test_branches.py (hunk 15 above)
Comment (pupper68k) [reply tests/easier68k/core/opcodes/test_branches.py]: again, intentional. we are using that test.

File: easier68k/core/opcodes/Bcc.py (hunk 11 above)
Comment (Chris-Johnston) [reply easier68k/core/opcodes/Bcc.py]: but isn't that exactly the condition you are checking on line 709?

File: easier68k/core/opcodes/Bcc.py (hunk 12 above)
Comment (Chris-Johnston) [reply easier68k/core/opcodes/Bcc.py]: Do you have the page number of where this is in the manual? I'm only able to find the checks on floating point numbers.

Two other sources I like to use are http://68k.hax.com/Bcc and http://wpage.unina.it/rcanonic/didattica/ce1/docs/68000.pdf because I find them much more readable. Both of these list `Z + N*!V + !N*V`

We can simplify the expression `N*!V + !N*V` to be `N != V`:
```
+---+---+-------------+
| N | V | N*!V + !N*V |
+---+---+-------------+
| 0 | 0 | 0           |
| 0 | 1 | 1           |
| 1 | 0 | 1           |
| 1 | 1 | 0           |
+---+---+-------------+
```

Checking for `not Z` after line 743 is redundant. We could actually reduce the whole thing to `return (n != v) or z`

```
+---+---+---+-----------------+
| N | V | Z | Z + N*!V + !N*V |
+---+---+---+-----------------+
| 0 | 0 | 0 | 0               |
| 0 | 0 | 1 | 1               |
| 0 | 1 | 0 | 1               |
| 0 | 1 | 1 | 1               |
| 1 | 0 | 0 | 1               |
| 1 | 0 | 1 | 1               |
| 1 | 1 | 0 | 0               |
| 1 | 1 | 1 | 1               |
+---+---+---+-----------------+
```

File: easier68k/core/opcodes/Bcc.py (hunk 13 above)
Comment (Chris-Johnston) [reply easier68k/core/opcodes/Bcc.py]: The test for LE is `Z or N and !V or !N and V`. This looks like the condition for GT.

File: easier68k/core/opcodes/Bcc.py [hunk 16]
```diff
@@ -0,0 +1,737 @@
...
+
+    @classmethod
+    def from_str(cls, command: str, parameters: str):
+        return bcc_from_str(command, parameters, BGT)
+
+
+class BLE(branch_code):
+    """
+    The BLE opcode compiles down to 6Fxx xxxx xxxx
+    The conditional here is (Z) OR (V AND NOT Z) OR (NOT N AND NOT V AND NOT Z)
```
Comment (Chris-Johnston) [inline easier68k/core/opcodes/Bcc.py]:
```suggestion
    The conditional here is (Z) OR (N AND NOT V) OR (NOT N AND V)
```

File: easier68k/core/opcodes/Bcc.py [hunk 17]
```diff
@@ -0,0 +1,737 @@
...
+    """
+
+    cond_code = '\x0F'
+
+    def conditional(self, simulator: M68K):
+        n = simulator.get_condition_status_code(ConditionStatusCode.N)
+        v = simulator.get_condition_status_code(ConditionStatusCode.V)
+        z = simulator.get_condition_status_code(ConditionStatusCode.Z)
+
+        return (n != v) and not z
```
Comment (Chris-Johnston) [inline easier68k/core/opcodes/Bcc.py]: The Z condition is not correct here.

```suggestion
        return (n != v) or z
```

File: easier68k/core/util/find_module.py [hunk 18]
```diff
@@ -11,7 +11,22 @@
     'easier68k.core.opcodes.lea.Lea',
     'easier68k.core.opcodes.trap.Trap',
     'easier68k.core.opcodes.opcode_or.Or',
-    'easier68k.core.opcodes.add.Add'
+    'easier68k.core.opcodes.add.Add',
+    'easier68k.core.opcodes.Bcc.BRA',
```
Comment (Chris-Johnston) [inline easier68k/core/util/find_module.py]: Sorry that this comment was made so late, but I think we should follow the pattern that filenames are lowercase, and classnames are CamelCase.

`opcode_or` was an exception because of the keyword `or`

File: tests/run_doctest.py [hunk 19]
```diff
@@ -32,6 +32,7 @@
     'easier68k.core.opcodes.neg',
     'easier68k.core.opcodes.simhalt',
     'easier68k.core.opcodes.trap',
+    'easier68k.core.opcodes.Bcc',
```
Comment (Chris-Johnston) [inline tests/run_doctest.py]: Same comment about filenames

File: easier68k/core/opcodes/bcc.py [hunk 20]
```diff
@@ -0,0 +1,737 @@
+# -*- coding: utf-8 -*-
+
+from ...core.enum.condition_status_code import ConditionStatusCode
+from ...simulator.m68k import M68K
+from ...core.opcodes.opcode import Opcode
+from ...core.util import opcode_util
+from ..util.parsing import *
+
+# makes TC way easier lol
```
Comment (Chris-Johnston) [inline easier68k/core/opcodes/bcc.py]: If you have suggestions on how we could improve type conversion overall (perhaps in the base opcode class as well), that would be super useful!

File: easier68k/core/opcodes/bcc.py [hunk 21]
```diff
@@ -0,0 +1,737 @@
...
+    '\x0F': 'BLE'
+}
+
+class branch_code(Opcode):
+    """
+    All Branch operations will inherit from branch_code.
+    there will be a conditional function to be overidden across child classes
+    one must also redefine self.cond_code across child classes
+    Additionally, a redefinition of command_matches(), from_str(), and disassemble_instruction() will suffice
+    """
```
Comment (Chris-Johnston) [inline easier68k/core/opcodes/bcc.py]: I think we should actually merge these two docstrings, since I think that our docs (which are WIP) would just ignore the second docstring block.

File: easier68k/core/opcodes/bcc.py [hunk 22]
```diff
@@ -0,0 +1,737 @@
...
+        # offset to bits
+        self.offset = Bits(int=self.offset,
+                           length=(self.size.get_number_of_bytes() * 8)).int
+
+        # Note on Page 130 of the NXP 68K Manual
+        # 0x00 8 bit offset defaults to word size
+        if self.offset == 0:
+            self.size = OpSize.WORD
+
+    def conditional(self, simulator: M68K):
```
Comment (Chris-Johnston) [inline easier68k/core/opcodes/bcc.py]: could we also add the [`@abstractmethod`](https://docs.python.org/3/library/abc.html#abc.abstractmethod) decorator to these functions as well?

File: easier68k/core/opcodes/bcc.py [hunk 23]
```diff
@@ -0,0 +1,737 @@
...
+        :param simulator:
+        :return: nothing
+        """
+
+        # check conditional
+        if self.conditional(simulator):
+            self.offset += 2
+            simulator.increment_program_counter(self.offset)
+
+        pass
```
Comment (Chris-Johnston) [inline easier68k/core/opcodes/bcc.py]:
```suggestion
```
Correct me if I'm wrong, but `pass` does nothing here?

File: easier68k/core/opcodes/bcc.py [hunk 24]
```diff
@@ -0,0 +1,737 @@
...
+        :return: string rep of command
+        """
+
+        return "{s} ${d}".format(
+                s = COND_CODE_TO_OPCODE.get(self.cond_code),
+                d = hex(self.operand or
+                     (self.address + (self.size if self.size.get_number_of_bytes() > 1 else 0) + self.offset)))
+
+    @classmethod
+    def command_matches(cls, command: str) -> bool:
```
Comment (Chris-Johnston) [inline easier68k/core/opcodes/bcc.py]: We could also use the `@abstractmethod` decorator here. I think the order matters, I used both of these in #132

File: easier68k/core/opcodes/bcc.py [hunk 25]
```diff
@@ -0,0 +1,737 @@
...
+        parameters[0] = parse_literal(parameters[0])
+        parameters[1] = int(parameters[1])
+
+        offset = abs(int(parameters[0]) - int(parameters[1] + 2))
+        size = 1 if offset < 128 else (2 if offset < 326778 else 3)
+
+        return size
+    
+    @classmethod
+    def is_valid(cls, command: str, parameters: str) -> (bool, list):
```
Comment (Chris-Johnston) [inline easier68k/core/opcodes/bcc.py]: These are fine without the decorator, since you don't just raise an exception or do nothing.

File: easier68k/core/opcodes/bcc.py [hunk 26]
```diff
@@ -0,0 +1,737 @@
...
+        # check offset dest validity (odd read)
+        operand = parse_literal(parameters[0].strip())
+        if operand % 2 == 1:
+            passes = False
+            errors.append("Reading word at {d} will cause crash".format(d=operand))
+            
+        return (passes, errors)
+    
+    @classmethod
+    def disassemble_instruction(cls, data: bytes) -> Opcode:
```
Comment (Chris-Johnston) [inline easier68k/core/opcodes/bcc.py]: `@abstractmethod`

File: easier68k/core/opcodes/bcc.py [hunk 27]
```diff
@@ -0,0 +1,737 @@
...
+        :param data: The data used to convert into an opcode instance
+        :return: The constructed instance or none if there was an error and
+            the amount of data in words that was used (e.g. extra for immediate
+            data) or 0 for not a match
+        """
+        raise NotImplementedError("Branch Operation did not override parent class disassemble_instruction stub!")
+
+
+    @classmethod
+    def from_str(cls, command: str, parameters: str):
```
Comment (Chris-Johnston) [inline easier68k/core/opcodes/bcc.py]: `@abstractmethod`

File: easier68k/core/opcodes/bcc.py (hunk 20 above)
Comment (pupper68k) [reply easier68k/core/opcodes/bcc.py]: TC here refers to Twos Compliment.

Comment (pupper68k) [reply easier68k/core/opcodes/bcc.py]: the Bitstring class will handle signs and casting for me. I find it super convenient.

Comment (Chris-Johnston) [reply easier68k/core/opcodes/bcc.py]: Oh, oops. Yeah 2s comp is used a lot, so maybe we could try to use this instead

File: easier68k/core/opcodes/#bcc.py# [hunk 28]
```diff
@@ -0,0 +1,740 @@
+# -*- coding: utf-8 -*-
+
```
Comment (Chris-Johnston) [inline easier68k/core/opcodes/#bcc.py#]: Not sure where #bcc.py# came from, don't think it should be here

File: easier68k/core/opcodes/bcc.py [hunk 29]
```diff
@@ -0,0 +1,737 @@
...
+        # offset to bits
+        self.offset = Bits(int=self.offset,
+                           length=(self.size.get_number_of_bytes() * 8)).int
+
+        # Note on Page 130 of the NXP 68K Manual
+        # 0x00 8 bit offset defaults to word size
+        if self.offset == 0:
+            self.size = OpSize.WORD
+
+;    @astractmethod
```
Comment (Chris-Johnston) [inline easier68k/core/opcodes/bcc.py]:
```suggestion
    @astractmethod
```
Found your build error

Comment (Chris-Johnston) [reply easier68k/core/opcodes/bcc.py]: if this depends on #132 , then let's just remove it from this PR and I'll do it in there
